src/main/scala/com/fulcrumgenomics/bam/pileup/RandomAccessPileupBuilder.scala

```diff
@@ -0,0 +1,94 @@
+    includeSupplementalAlignments: Boolean      = DefaultIncludeSupplementalAlignments,
+    includeMapPositionsOutsideFrInsert: Boolean = DefaultIncludeMapPositionsOutsideFrInsert,
+  ): RandomAccessPileupBuilder = {
+    require(source.indexed, "SAM source must be indexed for random access!")
+    new RandomAccessPileupBuilder(
+      source                             = source,
+      minMapQ                            = minMapQ,
+      minBaseQ                           = minBaseQ,
+      mappedPairsOnly                    = mappedPairsOnly,
+      includeDuplicates                  = includeDuplicates,
+      includeSecondaryAlignments         = includeSecondaryAlignments,
+      includeSupplementalAlignments      = includeSupplementalAlignments,
+      includeMapPositionsOutsideFrInsert = includeMapPositionsOutsideFrInsert,
+    )
+  }
+}
+
+/** A pileup builder that builds pileups using index-based BAM random access.
+  *
+  * @param source the indexed [[SamSource]] of records we will pileup.
+  * @param minBaseQ the minimum base quality that a base must have to contribute to a pileup.
+  * @param minMapQ the minimum mapping quality a record must have to contribute to a pileup.
+  * @param mappedPairsOnly if true, only allow paired records with both records mapped to contribute to a pileup.
+  * @param includeDuplicates if true, allow records flagged as duplicates to contribute to a pileup.
+  * @param includeSecondaryAlignments if true, allow records flagged as secondary alignments to contribute to a pileup.
+  * @param includeSupplementalAlignments if true, allow records flagged as supplementary alignments to contribute to a pileup.
+  * @param includeMapPositionsOutsideFrInsert if true, include any record of an FR pair where the site requested is outside the insert.
+  */
+class RandomAccessPileupBuilder private(
+  source: SamSource,
+  override val minMapQ: Int                                = DefaultMinMapQ,
+  override val minBaseQ: Int                               = DefaultMinBaseQ,
+  override val mappedPairsOnly: Boolean                    = DefaultMappedPairsOnly,
+  override val includeDuplicates: Boolean                  = DefaultIncludeDuplicates,
+  override val includeSecondaryAlignments: Boolean         = DefaultIncludeSecondaryAlignments,
+  override val includeSupplementalAlignments: Boolean      = DefaultIncludeSupplementalAlignments,
+  override val includeMapPositionsOutsideFrInsert: Boolean = DefaultIncludeMapPositionsOutsideFrInsert,
+) extends CloseablePileupBuilder {
+
+  /** The sequence dictionary associated with the records we will pileup. */
+  override val dict: SequenceDictionary = source.dict
+
+  /** Pileup records at this position. */
+  override def pileup(refName: String, pos: Int): Pileup[PileupEntry] = {
+    val end         = Math.min(pos + 1, dict(refName).length) // Add 1bp to "end" in case a read has an insertion there.
```

Comment:
This was an interesting find! Querying a 2bp region is required for correct behavior in returning records that start with insertion operators since an insertion entry is returned by `.query()` on the _next_ location whereas the pileup builder code puts insertions on the location prior to the insertion event.

Reply:
Great find!